client/protocols/analytics.py

```diff
@@ -21,6 +25,70 @@ def run(self, messages):
             # TODO(denero) Get the canonical name of the question
             statistics['question'] = self.args.question
 
+        statistics['started'] = self.check_start(messages['file_contents'])
+        print("DEBUG:" + statistics['started'])
+
         messages['analytics'] = statistics
 
+    def check_start(self, files):
+        """returns a dictionary where the key is question number, and the value
+        signals whether the question has been started.
+        """
+        begin = re.compile(r"""
+            [ \t]*
+            \#[ ]BEGIN[ ](.+)    # \1 is the question tag
+            """, re.X | re.M)
+        end = re.compile(r"""
+            [ \t]*
+            \#[ ]END[ ](.+)    # \1 is the question tag
+            """, re.X | re.M)
+        replace = re.compile(r"""
+            \#[ ]Replace[ ].+
+            """, re.X | re.M)
+
+        q_status = {}
+
+        for path in files:
+            lines = files[path].splitlines()
+            if len(lines) == 0:
+                log.warning("File {0} has no content".format(path))
+            line_num = 0
+            started = False
+            in_block = False
+            prev_begin_tag = None
+
```

Comment:
The text processing here can be made clearer and "simpler" by modifying the regular expressions to capture the content between the `BEGIN` and `END` tags:

```
tag = re.compile(r"""
\s+\#\s+BEGIN\s+(.+?)\n  # \1 is test name
\s+"*** REPLACE THIS LINE ***"\n
\s+\#\s+END\s+\1   # End tag
""", re.X | re.S)
```
(there's probably a few typos in my regex, you'll have to play around with it)

With this, `match.group(1)` is the test name and a match would only occur if it read
```
# BEGIN Test name
"*** REPLACE THIS LINE ***"
# END Test name
. Then you can check for started questions with something like the following:

```
# Initialize all tests to "started"
question_statuses = {test.name: True for test in self.assign.test_map.values()}

for path, file_contents in files.items():
    for test_name in tag.findall(file_contents):
        # Only matches tags for questions that' haven't been started
        question_status[test_name] = False
````